### ISSUE TITLE
Postgresql CTE recursive requires WITH RECURSIVE

### ISSUE THREAD

**Issue by hahn-kev** [linq2db/linq2db#1168]
Per PostgreSQL documentation it requires the `RECURSIVE` keyword when using a recursive CTE. 
https://www.postgresql.org/docs/current/static/queries-with.html

### Steps to reproduce

```c#
return _connection.GetCte<OrgGroup>(parents =>
{
    return (
            from orgGroup in OrgGroups
            where orgGroup.ParentId == null
            select orgGroup
        )
        .Union(
            from orgGroup in OrgGroups
            from parent in parents.InnerJoin(parent => parent.Id == orgGroup.ParentId)
            select orgGroup
        );
});
```

```sql
WITH parents
(
    "GroupName",
    "Type",
    "Supervisor",
    "ParentId",
    "ApproverIsSupervisor",
    "Id"
)
AS
(
  SELECT
    t1."GroupName",
    t1."Type" as "Type1",
    t1."Supervisor",
    t1."ParentId",
    t1."ApproverIsSupervisor",
    t1."Id"
  FROM
    "OrgGroup" t1
  WHERE
    t1."ParentId" IS NULL
  UNION
  SELECT
    t2."GroupName",
    t2."Type" as "Type1",
    t2."Supervisor",
    t2."ParentId",
    t2."ApproverIsSupervisor",
    t2."Id"
  FROM
    "OrgGroup" t2
    INNER JOIN parents parent ON parent."Id" = t2."ParentId"
)

SELECT
  t3."GroupName"            as "GroupName2",
  t3."Type"                 as "Type3",
  t3."Supervisor"           as "Supervisor2",
  t3."ParentId"             as "ParentId2",
  t3."ApproverIsSupervisor" as "ApproverIsSupervisor2",
  t3."Id"                   as "Id2"
FROM
  parents t3
```

### Exception
Npgsql.PostgresException Message  : 42P01: relation "parents" does not exist

Testing out this sql directly the error is inside the CTE, adding  `WITH RECURSIVE` fixes this error.

Taking a look at the code here
https://github.com/linq2db/linq2db/blob/master/Source/LinqToDB/SqlProvider/BasicSqlBuilder.cs#L335
If we would need to do something to add the `RECURSIVE` keyword if `with.IsRecursive` for PostgreSQL, but I don't see a good way to do that right now.

### Environment details
linq2db version: *2.0.0*  
Database Server: *E.g. PostgreSQL 9.5*  
Database Provider: *E.g. Npgsql 4*  
Operating system: *E.g. Windows 10*  
Framework version: *.NET Core 2.1*

**Comment by sdanyliv** [comment]
Thanks for reporting.
Better way for fix that is to add virtual property for BasicSqlBuilder - IsRecursiveCteKeywordRequired.
And render `RECURSIVE` keyword when it is needed.
Also we have to re-check which providers requires this keyword in recursive queries.

**Comment by sdanyliv** [comment]
Fixed in #1179
